tourney/player_skill.py

```diff
@@ -0,0 +1,176 @@
+
+  def save(self):
+    player_skills = [[p, ps.mu, ps.sigma] for p, ps in self.__player_skills.items()]
+    data = {
+      "player_skills": player_skills
+    }
+    os.makedirs(os.path.dirname(self.file_path()), exist_ok=True)
+    with open(self.file_path(), "w+") as fp:
+      json.dump(data, fp, indent=2)
+
+  def reset(self):
+    self.__player_skills = {}
+
+  def get_player_skill(self, user_id):
+    if user_id not in self.__player_skills:
+      self.__player_skills[user_id] = Rating()
+    return self.__player_skills[user_id]
+
+  def get_team_skill(self, team):
+    """ Average all mu values for the total team skill.
+    Lower confidence a lot, since we're not sure about this at all.
+    TODO: How much to lower the confidence?
+    """
+    team_skills = [self.get_player_skill(p) for p in team]
+    avg_mu = sum(r.mu for r in team_skills) / len(team)
+    # mu_dist = max([r.mu for r in team_skills]) - min([r.mu for r in team_skills])
+    avg_sigma = sum(r.sigma for r in team_skills) / len(team)
+    avg_sigma = avg_sigma * 2  # Increase uncertainty in averaged team skills
+    team_skill = Rating(mu=avg_mu, sigma=avg_sigma)
+    return team_skill
+
+  def get_match_quality(self, match):
+    team_a, team_b = match[0], match[1]
+    # Two teams in one match, number of rounds ignored
+    if len(match[0]) == len(match[1]):
+      team_a_skills = [self.get_player_skill(p) for p in team_a]
+      team_b_skills = [self.get_player_skill(p) for p in team_b]
+      return quality([team_a_skills, team_b_skills])
+    else:
+      # Unmatched teams, aggregate team skill and rate as 1vs1
+      skill_team_a = self.get_team_skill(team_a)
+      skill_team_b = self.get_team_skill(team_b)
+      return quality_1vs1(skill_team_a, skill_team_b)
+
+  def rate_uneven_match(self, win_team, lose_team):
+    """Rates a match between differently-sized teams implementation,
+    adjust each player as though they individually had won/lost
+    against the average of the opposing team.
+    TODO: Take into account that we might have really low confidence in the team.
+    """
+    win_team_skill = self.get_team_skill(win_team)
+    lose_team_skill = self.get_team_skill(lose_team)
+    new_win_team = []
+    new_lose_team = []
+    try:
+      new_win_team_skill, new_lose_team_skill =\
+        rate_1vs1(win_team_skill, lose_team_skill)
+
+      for p in win_team:
+        pskill = self.get_player_skill(p)
+        new_skill, _ = rate_1vs1(pskill, lose_team_skill)
+        self.__player_skills[p] = new_skill
+        new_win_team.append(new_skill)
+      for p in lose_team:
+        pskill = self.get_player_skill(p)
+        _, new_skill = rate_1vs1(win_team_skill, pskill)
+        self.__player_skills[p] = new_skill
+        new_lose_team.append(new_skill)
+    except FloatingPointError:
+      print("Skill difference too high: {} vs {}. Ratings not updated".
+            format(win_team_skill, lose_team_skill))
```

Comment:
Note that this will skip possible next participants as well. It would be best if it only skipped in problematic entity? In other words, maybe the `try` should be put directly on the `rate_1vs1()` invocations instead?

Reply:
Or wrap in a local wrapper lambda or function perhaps.